frame/benchmarking/src/lib.rs

```diff
@@ -903,6 +928,87 @@ macro_rules! impl_benchmark_test {
 	};
 }
 
+/// This creates a test suite which runs the module's benchmarks.
+///
+/// When called in [`pallet_example`] as
+///
+/// ```rust,ignore
+/// impl_benchmark_test_suite!(Module, crate::tests::new_test_ext, crate::tests::Test);
+/// ```
+///
+/// It expands to the equivalent of:
+///
+/// ```rust,ignore
+/// #[cfg(test)]
+/// mod tests {
+/// 	use super::*;
+/// 	use crate::tests::{new_test_ext, Test};
+/// 	use frame_support::assert_ok;
+///
+/// 	#[test]
+/// 	fn test_benchmarks() {
+/// 		new_test_ext().execute_with(|| {
+/// 			assert_ok!(test_benchmark_accumulate_dummy::<Test>());
+/// 			assert_ok!(test_benchmark_set_dummy::<Test>());
+/// 			assert_ok!(test_benchmark_another_set_dummy::<Test>());
+/// 			assert_ok!(test_benchmark_sort_vector::<Test>());
+/// 		});
+/// 	}
+/// }
+/// ```
+///
+/// ## Arguments
+///
+/// The first argument, `module`, must be the path to this crate's module.
+///
+/// The second argument, `new_test_ext`, must be the path to a function which takes no arguments
+/// and returns either a `sp_io::TestExternalities`, or some other type with an identical interface.
+///
+/// The third argument, `test`, must be the path to the runtime. The item to which this must refer
+/// will generally take the form:
+///
+/// ```rust,ignore
+/// frame_support::construct_runtime!(
+/// 	pub enum Test where ...
+/// 	{ ... }
+/// );
+/// ```
+///
+// ## Notes (not for rustdoc)
+//
+// The biggest challenge for this macro is communicating the actual test functions to be run. We
+// can't just build an array of function pointers to each test function and iterate over it, because
+// the test functions are parameterized by the `Test` type. That's incompatible with
+// monomorphization: if it were legal, then even if the compiler detected and monomorphized the
+// functions into only the types of the callers, which implementation would the function pointer
+// point to? There would need to be some kind of syntax for selecting the destination of the pointer
+// according to a generic argument, and in general it would be a huge mess and not worth it.
+//
+// Instead, we're going to steal a trick from `fn run_benchmark`: generate a function which is
+// itself parametrized by `Test`, which accepts a `&[u8]` parameter containing the name of the
+// benchmark, and dispatches based on that to the appropriate real test implementation. Then, we can
+// just iterate over the `Benchmarking::benchmarks` list to run the actual implementations.
+#[macro_export]
+macro_rules! impl_benchmark_test_suite {
+	($bench_module:tt, $new_test_ext:path, $test:path) => {
+		#[cfg(test)]
+		mod tests {
+			use super::{test_bench_by_name, $bench_module};
```

Comment:
`test_bench_by_name` is still used implicitly by macro

Reply:
Actually, I'm not sure that we want to make this explicit: the function is ultimately generated by the `benchmarks!` macro, and isn't visible to the user unless they do something like `cargo doc --all-features`. Passing it in explicitly gives the user more flexibility in terms of putting the `impl_benchmark_test_suite` invocation somewhere other than the `benchmarks` invocation, but if we require the user to provide it, it'll just be noise an easy majority of the time.

What do you think of this: we could add an optional parameter `$path_to_benchmarks_invocation:path` which defaults to `super`. We then `use  $path_to_benchmarks_invocation::test_bench_by_name;`. It's still kind of implicit, but when the user needs to deal with it, it's less abstruse; they just fill in the path to the module in which they invoked `benchmarks!`.

Reply:
another idea would be to move `impl_benchmark_test_suite` inside benchmark macro itself as an optional final argument.
I don't really now what is best

Reply:
ebf64f040 has what I was talking about. I think that's more usable than requiring the end-user to specify a generated function.